core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+    /// the data to the data file and then creating a new index
+    /// for it.
+    /// # Errors
+    /// Fails if any of the required platform specific function operations
+    /// fail.
+    fn append_block_to_chain(&mut self, block_data: &[u8]) -> Result<()> {
+        let new_block_height = self.read_index_count()?;
+        let start_location_in_data_file = if new_block_height == 0 {
+            0
+        } else {
+            let (ultimate_block_start, ultimate_block_len) =
+                self.read_block_index(new_block_height - 1)?;
+            ultimate_block_start + ultimate_block_len
+        };
+
+        self.write_block_data(start_location_in_data_file, block_data)?;
+        self.write_block_index(
+            new_block_height,
+            start_location_in_data_file,
+            block_data.len() as u64,
+        )?;
+
+        Ok(())
+    }
+}
+
+/// An implementation of a block store for kura
+/// that uses rusts default file io. `std::fs`.
+pub struct StdFileBlockStore {
+    path_to_blockchain: PathBuf,
+}
+
+impl StdFileBlockStore {
+    /// Create a new block store that stores blocks in the directory
+    /// pointed to by `path`.
+    pub fn new(path: &Path) -> Self {
+        StdFileBlockStore {
+            path_to_blockchain: path.to_path_buf(),
         }
     }
 }
 
-impl<IO: DiskIO> KuraWithIO<IO> {
-    /// After constructing [`Kura`] it should be initialized to be ready to work with it.
-    ///
-    /// # Errors
-    /// * May fail if file storage is either unavailable or data is invalid/corrupted
-    ///
-    #[iroha_futures::telemetry_future]
-    pub async fn init(&mut self) -> Result<Vec<VersionedCommittedBlock>> {
-        let blocks = self
-            .block_store
-            .read_all()
-            .await?
-            .try_collect::<Vec<_>>()
-            .await?;
-        self.merkle_tree = blocks
-            .iter()
-            .map(VersionedCommittedBlock::hash)
-            .collect::<MerkleTree<_>>();
-        Ok(blocks)
+impl BlockStoreTrait for StdFileBlockStore {
+    #[allow(
+        clippy::needless_range_loop,
+        clippy::unwrap_used,
+        clippy::unwrap_in_result
+    )]
+    fn read_block_indices(
+        &self,
+        start_block_height: u64,
+        dest_buffer: &mut [(u64, u64)],
+    ) -> Result<()> {
+        let mut index_file = std::fs::OpenOptions::new()
+            .read(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let start_location = start_block_height * 2 * 8;
+        let block_count = dest_buffer.len() as u64;
+
+        if start_location + (2 * 8) * block_count > index_file.metadata()?.len() {
+            return Err(Error::OutOfBoundsBlockRead(start_block_height, block_count));
+        }
+        index_file.seek(SeekFrom::Start(start_location))?;
+        // (start, length), (start,length) ...
+        for i in 0..block_count {
+            let index: usize = i.try_into().unwrap();
```

Comment:
So if this is an error that should come up more frequently, I'm not a huge fan of `unwrap`-ing. 
- `expect` also panics, but gives you an error message, which I think would be appropriate here. 
- This is a potentially recoverable error, which only occurs if `u64 > usize`, i.e. a 32-bit architecture, where we should have a workaround. That is, you can have another array of chunks, which can then be subscripted by the 32-bit sub-words of a single `u64`.

Reply:
If the index fits in 32 bits then it doesn't fail. When `i >= 2^32` it will panic, and that makes sense because a lot of code has to change to load and parse a 4 billion block chain on a 32 bit machine. It can be done but then most of this code must be redone.

Reply:
I think we've had an agreement to always prefer `expect` vs `unwrap`.

Reply:
I'll change them to expects to honor such an agreement. But I think I end up repeating myself. The code already makes it obvious what has gone wrong and then I type it out again. Seems foolish.